src/ros_filter.cpp

```diff
@@ -2180,6 +2057,164 @@ void RosFilter<T>::periodicUpdate()
+{
+  updateWorldToBaseLinkTransform(filtered_position);
+
+  // The filtered_position is the message containing the state and covariances:
+  // nav_msgs Odometry
+  if (!validateFilterOutput(filtered_position)) {
+    RCLCPP_ERROR(
+      this->get_logger(),
+      "Critical Error, NaNs were detected in the output state of the filter. "
+      "This was likely due to poorly conditioned process, noise, or sensor "
+      "covariances.");
+  }
+
+  // If we're trying to publish with the same time stamp, it means that we had a measurement get
+  // inserted into the filter history, and our state estimate was updated after it was already
+  // published. As of ROS Noetic, TF2 will issue warnings whenever this occurs, so we make this
+  // behavior optional. Just for safety, we also check for the condition where the last published
+  // stamp is *later* than this stamp. This should never happen, but we should handle the case
+  // anyway.
+  bool corrected_data = (!permit_corrected_publication_ &&
+    last_published_stamp_ >= filtered_position.header.stamp);
+
+  if (publish_transform_ && !corrected_data) {
+    publishWorldTransform(filtered_position);
+  }
+
+  // Retain the last published stamp so we can detect repeated transforms in future cycles
+  last_published_stamp_ = filtered_position.header.stamp;
+
+  // Fire off the position and the transform
+  if (!corrected_data) {
+    position_pub_->publish(std::move(filtered_position));
+  }
+
+  if (print_diagnostics_) {
+    freq_diag_->tick();
+  }
+
+  return corrected_data;
+}
+
+template<typename T>
+void RosFilter<T>::updateWorldToBaseLinkTransform(const nav_msgs::msg::Odometry & filtered_position)
+{
+  world_base_link_trans_msg_.header.stamp =
+    static_cast<rclcpp::Time>(filtered_position.header.stamp) + tf_time_offset_;
+  world_base_link_trans_msg_.header.frame_id =
+    filtered_position.header.frame_id;
+  world_base_link_trans_msg_.child_frame_id =
+    filtered_position.child_frame_id;
+
+  world_base_link_trans_msg_.transform.translation.x =
+    filtered_position.pose.pose.position.x;
+  world_base_link_trans_msg_.transform.translation.y =
+    filtered_position.pose.pose.position.y;
+  world_base_link_trans_msg_.transform.translation.z =
+    filtered_position.pose.pose.position.z;
+  world_base_link_trans_msg_.transform.rotation =
+    filtered_position.pose.pose.orientation;
+}
+
+template<typename T>
+void RosFilter<T>::publishWorldTransform(const nav_msgs::msg::Odometry & filtered_position)
+{
+  // If the world_frame_id_ is the odom_frame_id_ frame, then we can just
+  // send the transform. If the world_frame_id_ is the map_frame_id_ frame,
+  // we'll have some work to do.
+
+  auto frame_id = filtered_position.header.frame_id;
+
+  if (frame_id == odom_frame_id_) {
+    world_transform_broadcaster_->sendTransform(world_base_link_trans_msg_);
+    return;
+  }
+
+  if (frame_id != map_frame_id_) {
+    std::cerr << "Odometry message frame_id was " <<
```

Comment:
As long as you're here, you can replace any `cerr` statements with the actual ROS 2 logging macros, and uncomment/update the ROS 1 macros further down.

Reply:
Fixed.